---
output: github_document
---
```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-"
)
```

# ggPlantmap <img src="man/figures/hex.png" align="right" height="200"/>
###### Version Beta.1

<!-- badges: start -->
<!-- badges: end -->

## Overview

ggPlantmap is a R tidyverse based package with a series of plant images to be mapped in a ggplot enviroment. We also provide a series of functions and instructions to easily create your own personal ggPlantmap. These ggPlantmap objects can be added into R pipelines for the visual representation of quantitative data in distinct plant cells and/or structures. 
<img src="man/figures/ggPlantmap.example1.jpg" align="center" height="250"/>

## Installation

```{r, eval = FALSE}
##install devtools (if you haven't already)
install.packages("devtools")
library(devtools)

## Installing from a github respository
install_github("leonardojo/ggPlantmap")
```


## What is a ggPlantmap?

Each unique ggPlantmap is a table (tibble) object with points coordinates (x,y) of specific polygons extracted from plant images.
```{r}
library(ggPlantmap)
head(ggPm.At.roottip.longitudinal)
```

## Where can I find the list of all ggPlantmaps objects?

The whole list of pre-loaded ggPlantmap objects can be found in the table ggPm.summary.
You can find the description of ggPlantmaps, as well as information of its creator. 
Because most ggPlantmaps are based on previously published plant images, the references for each specific image can also be found in the summary table. We hope to keep updating the ggPlantmap catalog, with the help of the plant research community.
<img src="man/figures/ggPm.someexamples.jpg" align="center" height="500"/></a>

```{r}
head(ggPm.summary)

##Listing all the ggPlantmap objects
ggPm.summary$ggPlantmap.name
```

## General usage

All ggPlantmaps are pre-loaded in the package, you can call them directly in your R environment by typing their name. 
```{r}
library(ggPlantmap)
##examples
ggPm.At.roottip.longitudinal
ggPm.At.roottip.crosssection
ggPm.Ms.root.crosssection
```

## How can I plot a ggPlantmap?

You can use the ggPlantmap.plot() function to quickly visualize your ggPlantmap.

```{r,warning=F,out.width="70%"}
##ggPlantmap.plot(data,layer,linewidth=0.5,show.legend=T)
ggPlantmap.plot(ggPm.At.roottip.longitudinal,ROI.id,linewidth = 1,show.legend = F)
```

If you have experience with ggplot, you can feed your a ggPlantmap object into a ggplot with the geom_polygon() function. 
```{r,warning=F,out.width="50%"}
library(ggplot2)
ggplot(ggPm.At.roottip.longitudinal,aes(x,y)) +
  geom_polygon(aes(group=ROI.id,fill=factor(ROI.id)),show.legend = F,colour="black",linewidth=1) +
  coord_fixed() ## important to keep the aspect ratio of the plot

```

## How can I colormap distinct layers of a ggPlantmap?

Because each polygon on ggPlantmap is characterized by specific levels (examples: Region,Stage,Part), you can color map them individually.
Just specify the column  you want to colormap in the layer option of the function.
ggPlantmap.plot() is based on ggplot so you can add specific modifications to it using ggplot coding logic.

```{r,warning=F}
library(ggplot2)
library(cowplot)

head(ggPm.At.seed.devseries)

## Stage: Seed development stage
## Part: Distinct parts of a seed (Seed coat, Endosperm and Embryo)
## Region: Specific regions of each part of the Arabidopsis seed
## Reference: Belmonte, Mark F., et al. "Comprehensive developmental profiles of gene activity in regions and subregions of the Arabidopsis seed." Proceedings of the National Academy of Sciences 110.5 (2013): E435-E444.

a <- ggPlantmap.plot(ggPm.At.seed.devseries,Region,linewidth = 0.5) +
  scale_fill_brewer(palette="Set3") +
  ggtitle("Regions of Arabidopsis seed development") +
   theme(legend.key.height= unit(0.25, 'cm'),
        legend.key.width= unit(0.25, 'cm'))
b <- ggPlantmap.plot(ggPm.At.seed.devseries,Stage,linewidth = 0.5) +
  scale_fill_brewer(palette="Set1") +
  ggtitle("Stages of Arabidopsis seed development") +
   theme(legend.key.height= unit(0.25, 'cm'),
        legend.key.width= unit(0.25, 'cm'))
c <- ggPlantmap.plot(ggPm.At.seed.devseries,Part,linewidth = 0.5) +
  scale_fill_brewer(palette="Set1") +
  ggtitle("Parts of Arabidopsis seed development") +
   theme(legend.key.height= unit(0.25, 'cm'),
        legend.key.width= unit(0.25, 'cm'))
plot_grid(a,b,c,ncol=1,labels=c("a","b","c"),align = "v")
```

Each map will have their own classification. If you would like to adjust or create your own classification, you can save the ggPlantmap as a table and modify it on to mach the degree of separation you want to show.




## How can I create my own ggPlantmap?

The principle of creating a ggPlantmap is fairly simple. We generate a list of ROIs (region of interests) in the Icy open-source software (https://icy.bioimageanalysis.org/) from any image. These ROIs are saved as XML files and later be converted into ggPlantmaps by using the function XML.to.ggPlantmap() function. We created step-by-step tutorial on how to generate xml images from plant images, you can find the tutorial here. 

```{r}
##converting the sample file: ggPm.sample.xml into a ggPlantmap table
ggPm <- XML.to.ggPlantmap("data/ggPm.sample.xml")
head(ggPm)
##plotting the ggPm
ggPlantmap.plot(ggPm)

```

## How can I overlay quantitative data into my ggPlantmap?

Over the recent years, we are seeing a crescent interest on the characterization of molecular events that occur in specific cells/parts of the plant, such as single-cell sequencing approaches (ScRNA-seq,ScATAC-seq,TRAP-seq,LCM-RNAseq,etc) as well as high-resolution spatial profiling of RNAs in plant cells (PHYTOmap,In situ hybridization chain reaction (HCR), merFISH, Stereo-Seq,etc). These techniques offer powerful insights to understand cell-type specific events in a complex plant tissue. To better explore this type of data, it would be important to create tools that allow us to visualize and communicate the quantitative features of cells/parts the plant.

With ggPlantmap, we can overlay external quantitative data in the map as a form of a heatmap. You can combine the ggPlantmap with a external quantitative data using the ggPlantmap.merge() function. later, the heatmap can be generated using the ggPlantmap.heatmap() function. 

This approach can be very helpful for R Shiny app developers to create web interactive tools to visualize gene expression gene profiles.

 
```{r}
##Quantitative sample data, the expression of SCR in distinct cell-types of the Tomato root.
## Adapted data from: Kajala, Kaisa, et al. "Innovation, conservation, and repurposing of gene function in root cell type development." Cell 184.12 (2021): 3333-3348.
head(ggPm.tomatoatlas.expression.sample)
##important: Names in the quantitative data needs to match the ones found in the map.

##Merging both datasets
expression.sample2 <- ggPlantmap.merge(ggPm.Sl.root.crosssection,ggPm.tomatoatlas.expression.sample,id.x = "ROI.name",id.y="Cell.layer") ##Column names between tables are different, need to specify both identifiers in x and y.
head(expression.sample2)

##Ploting
ggPlantmap.heatmap(expression.sample2,SCR.expression) +
  scale_fill_gradient(low="white",high="Red",na.value ="white")
```


```{r}
##Quantitative sample data, the expression of AT5G47670 (LEC1-like) in distinct stages and parts of the Arabidopsis seed
## Data from: Belmonte, Mark F., et al. "Comprehensive developmental profiles of gene activity in regions and subregions of the Arabidopsis seed." Proceedings of the National Academy of Sciences 110.5 (2013): E435-E444.
head(ggPm.At.seed.expression.sample)
##important: Names in the quantitative data needs to match the ones found in the map.

##Merging both datasets
expression.sample <- ggPlantmap.merge(ggPm.At.seed.devseries,ggPm.At.seed.expression.sample,"ROI.name")
head(expression.sample)

##Ploting
ggPlantmap.heatmap(expression.sample,AT5G47670.expression) +
  scale_fill_gradient(low="white",high="Red")

```
 

## Is ggPlantmap only usefull for molecular expression data?

Not at all. ggPlantmap can also be used to produce many other type of plots. Essentially anything that you can trace, you can create! Be creative! We hope to build a community where people explore the usage of ggPlantmap for the communication of Plant science. 

<img src="man/figures/ggPm.otherexamples.png" align="center" height="500"/></a>

## Can my ggPlantmap be included in the package?

YES!!! Any Plant map can be included in the package. If you create one, please email me (l.jo@uu.nl) your ggPlantmap as tab-delimited table and I'll make sure to include in the package. You will be credited and your information will be displayed in the summary file. I really hope this becomes an organic package with the contribution of the plant research community.


## Ackowledgements

Soon
